In [245]:
import os
import pandas as pd
import math

import plotly.express as px
import matplotlib.pyplot as plt
import seaborn as sns

from collections import Counter

# Reading the data

In [157]:
fp = os.path.join(os.getcwd(),'Question2.xlsx')
fp

'd:\\Data\\OfficeWorkspace-20191016T044923Z-001\\OfficeWorkspace\\python_challenges\\nexera\\Problems\\Problems\\Question2.xlsx'

In [158]:
df = pd.read_excel(fp)\
    .fillna(0)

In [159]:
df=df[['Material', 'Shape', 'Color', 'Radius', 'Height',
       'Units (Imperial/Metric)', 'Pressure P1', 'Pressure P2',
       'Piston movement']]
df = df.replace('Up',1)
df = df.replace('Down',-1)
df

Material     Shape   Color    Radius    Height   
0   STAINLESS STEEL      CONE  orange  0.027432  0.024384  \
1          ALUMINUM    SPHERE    blue  0.090000  0.000000   
2       POLYSTYRENE      CUBE   black  0.000000  0.003048   
3             PAPER      CONE     red  0.040000  0.030000   
4       POLYSTYRENE      CUBE  orange  0.000000  0.060000   
..              ...       ...     ...       ...       ...   
95  STAINLESS STEEL  CYLINDER  orange  0.050000  0.100000   
96         ALUMINUM      CONE   white  0.080000  0.080000   
97              OAK  CYLINDER  orange  0.040000  0.020000   
98         ALUMINUM      CONE    blue  0.060000  0.020000   
99      POLYSTYRENE      CUBE   white  0.000000  0.040000   

   Units (Imperial/Metric)  Pressure P1  Pressure P2  Piston movement  
0                        I        95959       102836                1  
1                        M       100849       105818               -1  
2                        I        93328       106262                1  
3                        M        92008        96850                1  
4                        M       100407       106656                1  
..                     ...          ...          ...              ...  
95                       M       106221       104604               -1  
96                       M       107102       100344               -1  
97                       M       102819       107796                1  
98                       M       104587       103270               -1  
99                       M        97513        99863                1  

[100 rows x 9 columns]

In [160]:


# df3=df[[ 'Radius', 'Height','Pressure P1', 'Pressure P2',
#        'Piston movement']]
# corr = plt.matshow(df3.corr())

# corr_pairs = df3.corr().unstack()
# print(corr_pairs)

In [161]:
materials = ['STAINLESS STEEL', 'ALUMINUM', 'POLYSTYRENE', 'PAPER', 'OAK']
shapes = ['CONE', 'SPHERE', 'CUBE', 'CYLINDER']
colours = ['orange', 'blue', 'black', 'red', 'white', 'yellow']
df0 = df.copy()

In [162]:
def ismatch(x,str_):
    return 1 if x==str_ else 0

In [163]:
# restructuring data


df0['IMP'] = df['Units (Imperial/Metric)'].apply(ismatch,args={'I'})
df0['MET'] = df['Units (Imperial/Metric)'].apply(ismatch,args={'M'})
for s in materials:
    s = s.upper()
    df0[s] = df0['Material'].apply(ismatch,args={s})

for s in shapes:
    s = s.upper()
    df0[s] = df0['Shape'].apply(ismatch,args={s})

for col in colours:
    df0[col] = df0['Color'].apply(ismatch,args={col})




In [164]:
df00 = df0[['Radius', 'Height','Pressure P1', 'Pressure P2',
       'Piston movement','ALUMINUM', 'POLYSTYRENE', 'PAPER', 'OAK',
       'CONE', 'SPHERE', 'CUBE', 'CYLINDER', 'IMP', 'MET', 'orange', 'blue',
       'black', 'red', 'white', 'yellow', 'STAINLESS STEEL']]
df0

Material     Shape   Color    Radius    Height   
0   STAINLESS STEEL      CONE  orange  0.027432  0.024384  \
1          ALUMINUM    SPHERE    blue  0.090000  0.000000   
2       POLYSTYRENE      CUBE   black  0.000000  0.003048   
3             PAPER      CONE     red  0.040000  0.030000   
4       POLYSTYRENE      CUBE  orange  0.000000  0.060000   
..              ...       ...     ...       ...       ...   
95  STAINLESS STEEL  CYLINDER  orange  0.050000  0.100000   
96         ALUMINUM      CONE   white  0.080000  0.080000   
97              OAK  CYLINDER  orange  0.040000  0.020000   
98         ALUMINUM      CONE    blue  0.060000  0.020000   
99      POLYSTYRENE      CUBE   white  0.000000  0.040000   

   Units (Imperial/Metric)  Pressure P1  Pressure P2  Piston movement  IMP   
0                        I        95959       102836                1    1  \
1                        M       100849       105818               -1    0   
2                        I        93328       106262                1    1   
3                        M        92008        96850                1    0   
4                        M       100407       106656                1    0   
..                     ...          ...          ...              ...  ...   
95                       M       106221       104604               -1    0   
96                       M       107102       100344               -1    0   
97                       M       102819       107796                1    0   
98                       M       104587       103270               -1    0   
99                       M        97513        99863                1    0   

    ...  CONE  SPHERE  CUBE  CYLINDER  orange  blue  black  red  white  yellow  
0   ...     1       0     0         0       1     0      0    0      0       0  
1   ...     0       1     0         0       0     1      0    0      0       0  
2   ...     0       0     1         0       0     0      1    0      0       0  
3   ...     1       0     0         0       0     0      0    1      0       0  
4   ...     0       0     1         0       1     0      0    0      0       0  
..  ...   ...     ...   ...       ...     ...   ...    ...  ...    ...     ...  
95  ...     0       0     0         1       1     0      0    0      0       0  
96  ...     1       0     0         0       0     0      0    0      1       0  
97  ...     0       0     0         1       1     0      0    0      0       0  
98  ...     1       0     0         0       0     1      0    0      0       0  
99  ...     0       0     1         0       0     0      0    0      1       0  

[100 rows x 26 columns]

# Questions

### Q1 .Which variable seems to have the least impact on the piston movement?

#### correlation factors

In [173]:
corr_pairs = df00.corr().unstack()
corr_pairs['Piston movement']

Radius            -0.326952
Height            -0.217908
Pressure P1       -0.405633
Pressure P2        0.317644
Piston movement    1.000000
ALUMINUM           0.023637
POLYSTYRENE       -0.066722
PAPER              0.081747
OAK                0.136939
CONE              -0.018020
SPHERE            -0.073103
CUBE               0.143428
CYLINDER          -0.066722
IMP                0.230859
MET               -0.230859
orange             0.006063
blue               0.021194
black             -0.032642
red                0.021194
white              0.061338
yellow            -0.073103
STAINLESS STEEL   -0.108521
dtype: float64

##### normalizing data

In [178]:
df00_n =  df00.copy()

In [182]:
df00_n['norm Pressure P1'] = (df00_n['Pressure P1']-df00_n['Pressure P1'].mean())/df00_n['Pressure P1'].std()
df00_n['norm Pressure P2'] = (df00_n['Pressure P2']-df00_n['Pressure P2'].mean())/df00_n['Pressure P2'].std()

In [186]:
df00_n = df00_n[['Radius', 'Height','Piston movement',
       'ALUMINUM', 'POLYSTYRENE', 'PAPER', 'OAK', 'CONE', 'SPHERE', 'CUBE',
       'CYLINDER', 'IMP', 'MET', 'orange', 'blue', 'black', 'red', 'white',
       'yellow', 'STAINLESS STEEL', 'norm Pressure P1', 'norm Pressure P2']]

In [187]:
corr_pairs_n = df00_n.corr().unstack()
corr_pairs_n['Piston movement']

Radius             -0.326952
Height             -0.217908
Piston movement     1.000000
ALUMINUM            0.023637
POLYSTYRENE        -0.066722
PAPER               0.081747
OAK                 0.136939
CONE               -0.018020
SPHERE             -0.073103
CUBE                0.143428
CYLINDER           -0.066722
IMP                 0.230859
MET                -0.230859
orange              0.006063
blue                0.021194
black              -0.032642
red                 0.021194
white               0.061338
yellow             -0.073103
STAINLESS STEEL    -0.108521
norm Pressure P1   -0.405633
norm Pressure P2    0.317644
dtype: float64

#### A1. The "colour" of the object seems to be of the least significance pertinent to piston movement. We can see the correlation of all the columns wrt piston movement for raw data and the normalized data [for P1 and P2] and they show the same results

## Q2. Which shape seems to be the most common for red objects?

In [197]:
Counter(df[df.Color=='red'].Shape.tolist()).most_common()

[('CYLINDER', 5), ('CUBE', 4), ('CONE', 3), ('SPHERE', 3)]

#### A2. As can be seen from the counter above , Cylinder seems to be most common object 
of red colour

## Q3. What are the most common dimensions (or dimension range) for cubical objects?

In [202]:

df_cube = df[df.Shape=='CUBE']
df_cube

Material Shape   Color  Radius    Height Units (Imperial/Metric)   
2       POLYSTYRENE  CUBE   black     0.0  0.003048                       I  \
4       POLYSTYRENE  CUBE  orange     0.0  0.060000                       M   
5       POLYSTYRENE  CUBE     red     0.0  0.009144                       I   
9       POLYSTYRENE  CUBE    blue     0.0  0.040000                       M   
10  STAINLESS STEEL  CUBE  orange     0.0  0.050000                       M   
12            PAPER  CUBE  yellow     0.0  0.006096                       I   
17  STAINLESS STEEL  CUBE   black     0.0  0.003048                       I   
24      POLYSTYRENE  CUBE   white     0.0  0.020000                       M   
25              OAK  CUBE   black     0.0  0.100000                       M   
27         ALUMINUM  CUBE     red     0.0  0.070000                       M   
28  STAINLESS STEEL  CUBE   black     0.0  0.040000                       M   
29  STAINLESS STEEL  CUBE   black     0.0  0.003048                       I   
32            PAPER  CUBE    blue     0.0  0.010000                       M   
34      POLYSTYRENE  CUBE   black     0.0  0.070000                       M   
36         ALUMINUM  CUBE  orange     0.0  0.060000                       M   
39         ALUMINUM  CUBE  orange     0.0  0.080000                       M   
41      POLYSTYRENE  CUBE  yellow     0.0  0.090000                       M   
42  STAINLESS STEEL  CUBE   black     0.0  0.018288                       I   
44            PAPER  CUBE     red     0.0  0.021336                       I   
49  STAINLESS STEEL  CUBE   black     0.0  0.080000                       M   
52              OAK  CUBE  yellow     0.0  0.080000                       M   
59      POLYSTYRENE  CUBE   white     0.0  0.030000                       M   
69            PAPER  CUBE   white     0.0  0.090000                       M   
72         ALUMINUM  CUBE     red     0.0  0.060000                       M   
74      POLYSTYRENE  CUBE   black     0.0  0.020000                       M   
86  STAINLESS STEEL  CUBE  orange     0.0  0.015240                       I   
87  STAINLESS STEEL  CUBE    blue     0.0  0.020000                       M   
89      POLYSTYRENE  CUBE   black     0.0  0.080000                       M   
99      POLYSTYRENE  CUBE   white     0.0  0.040000                       M   

    Pressure P1  Pressure P2  Piston movement  
2         93328       106262                1  
4        100407       106656                1  
5        105055        96833               -1  
9        106568        98844               -1  
10        95343        98121               -1  
12       103594       103441               -1  
17        93161        99214                1  
24       109373        93090               -1  
25       100160        99668               -1  
27        94231        98119               -1  
28       105593       103914               -1  
29       101492        93867               -1  
32        96404       101749                1  
34       109139       108781               -1  
36       109412       103569               -1  
39       109811        93767               -1  
41       101129       109930                1  
42       101576       106110                1  
44       104183       106896                1  
49        95949       103049               -1  
52       111367        93435               -1  
59       105602        96600               -1  
69        96679       103124                1  
72        99068       111869                1  
74       103645       102453               -1  
86        92122        98867                1  
87       100858       109187                1  
89        99428        98153               -1  
99        97513        99863                1

In [203]:
df_cube.describe()

Radius     Height    Pressure P1    Pressure P2  Piston movement
count    29.0  29.000000      29.000000      29.000000        29.000000
mean      0.0   0.043767  101454.827586  101566.586207        -0.172414
std       0.0   0.031350    5548.759862    5240.614574         1.002460
min       0.0   0.003048   92122.000000   93090.000000        -1.000000
25%       0.0   0.018288   96679.000000   98121.000000        -1.000000
50%       0.0   0.040000  101129.000000  101749.000000        -1.000000
75%       0.0   0.070000  105593.000000  106110.000000         1.000000
max       0.0   0.100000  111367.000000  111869.000000         1.000000

In [240]:
height_range  = (df_cube.Height.mean() - df_cube.Height.std() ,  df_cube.Height.mean() + df_cube.Height.std())
height_range

(0.012417085690207247, 0.07511725913737897)

#### A3. from the above table we can see most of the cube shaped objects lie in the range of (0.012417085690207247, 0.07511725913737897)

## Q4. What is the most common characteristic among the top 10% hardest-to-manipulate objects?

### what is the definition of hardest to manipulate objects? 
Assumption 1 : is it all the objects that could not be moved up?

Assumption 2: is it the heaviest objects that could not be moved up?

Assumption 3: is it the most upward pressure applied for all the objects that could be moved up?

Assumption 4: is it the most  delta pressure applied for all the objects that could be moved up?
    
    - Assumption 4 makes the most sense and is also easy to verify so i will use this




In [252]:
df['delta_p'] = df['Pressure P2'] - df['Pressure P1']
df

Material     Shape   Color    Radius    Height   
0   STAINLESS STEEL      CONE  orange  0.027432  0.024384  \
1          ALUMINUM    SPHERE    blue  0.090000  0.000000   
2       POLYSTYRENE      CUBE   black  0.000000  0.003048   
3             PAPER      CONE     red  0.040000  0.030000   
4       POLYSTYRENE      CUBE  orange  0.000000  0.060000   
..              ...       ...     ...       ...       ...   
95  STAINLESS STEEL  CYLINDER  orange  0.050000  0.100000   
96         ALUMINUM      CONE   white  0.080000  0.080000   
97              OAK  CYLINDER  orange  0.040000  0.020000   
98         ALUMINUM      CONE    blue  0.060000  0.020000   
99      POLYSTYRENE      CUBE   white  0.000000  0.040000   

   Units (Imperial/Metric)  Pressure P1  Pressure P2  Piston movement  delta_p  
0                        I        95959       102836                1     6877  
1                        M       100849       105818               -1     4969  
2                        I        93328       106262                1    12934  
3                        M        92008        96850                1     4842  
4                        M       100407       106656                1     6249  
..                     ...          ...          ...              ...      ...  
95                       M       106221       104604               -1    -1617  
96                       M       107102       100344               -1    -6758  
97                       M       102819       107796                1     4977  
98                       M       104587       103270               -1    -1317  
99                       M        97513        99863                1     2350  

[100 rows x 10 columns]

In [260]:

df_hardest_to_move = df[df['Piston movement']==1]
df_hardest_to_move = df_hardest_to_move.sort_values(by='delta_p',ascending=False)
df_hardest_to_move_top_10 = df_hardest_to_move.head(10).reset_index(drop=True)
df_hardest_to_move_top_10

Material     Shape   Color    Radius    Height   
0  STAINLESS STEEL  CYLINDER  orange  0.020000  0.080000  \
1      POLYSTYRENE      CUBE   black  0.000000  0.003048   
2         ALUMINUM      CUBE     red  0.000000  0.060000   
3      POLYSTYRENE    SPHERE    blue  0.050000  0.000000   
4            PAPER      CONE   black  0.030000  0.040000   
5         ALUMINUM    SPHERE    blue  0.040000  0.000000   
6      POLYSTYRENE      CUBE  yellow  0.000000  0.090000   
7  STAINLESS STEEL      CUBE    blue  0.000000  0.020000   
8         ALUMINUM    SPHERE    blue  0.015240  0.000000   
9  STAINLESS STEEL      CONE  orange  0.027432  0.024384   

  Units (Imperial/Metric)  Pressure P1  Pressure P2  Piston movement  delta_p  
0                       M        92331       106933                1    14602  
1                       I        93328       106262                1    12934  
2                       M        99068       111869                1    12801  
3                       M        97146       107586                1    10440  
4                       M        95450       105481                1    10031  
5                       M        95872       104725                1     8853  
6                       M       101129       109930                1     8801  
7                       M       100858       109187                1     8329  
8                       I        98423       106001                1     7578  
9                       I        95959       102836                1     6877

In [263]:
shape_wise = Counter(df_hardest_to_move_top_10.Shape.tolist()).most_common()
print(f"shape_wise counter of hardest to move objects: {shape_wise}")
material_wise = Counter(df_hardest_to_move_top_10.Material.tolist()).most_common()
print(f"material_wise counter of hardest to move objects: {material_wise}")

shape_wise counter of hardest to move objects: [('CUBE', 4), ('SPHERE', 3), ('CONE', 2), ('CYLINDER', 1)]
material_wise counter of hardest to move objects: [('STAINLESS STEEL', 3), ('POLYSTYRENE', 3), ('ALUMINUM', 3), ('PAPER', 1)]


#### A4.

shape_wise counter of hardest to move objects: [('CUBE', 4), ('SPHERE', 3), ('CONE', 2), ('CYLINDER', 1)]

material_wise counter of hardest to move objects: [('STAINLESS STEEL', 3), ('POLYSTYRENE', 3), ('ALUMINUM', 3), ('PAPER', 1)]

Given the observations from above it looks like the material wise behaviour is inconclusive except that paper as  material  
is difficult to move only for the conical shape which makes sense. 

Hoever, looking at the shapes, it is clear that the cube shape offers the most challenge while being manipulated closely followed by the sphere.

If we now take the hardest to manipulate shapes and look at the involved materials, we can see that polystyrene is the material thats hardest to manipulate.


## Q5. Which row in the excel sheet corresponds to the largest object being manipulated?

In [274]:
vols={
    'CUBE': lambda radius,height: height ** 3,
    'SPHERE': lambda radius,height : 4.0 / 3.0 * math.pi * (radius ** 3),
    'CYLINDER': lambda radius,height : math.pi * (radius ** 2) * height,
    'CONE': lambda radius,height : math.pi * (radius ** 2) * height / 3.0,
}

In [285]:
def calc_vol(shape_, radius,height, vols=vols):
    
    return vols[shape_](radius,height)

calc_vol('CUBE',2,2)


8

In [289]:
vol_shape = []
for i in df.index:
    # print(i)
    shape__ = df.iloc[i].Shape
    radius__ = df.iloc[i].Radius
    height__ = df.iloc[i].Height

    vol = calc_vol(shape__, radius__, height__)
    # print('\t',vol)
    vol_shape.append(vol)

In [293]:
df['Volume'] = vol_shape

In [302]:
df_largest_vol = df.sort_values(by='Volume',ascending=False)
df_largest_vol

Material     Shape   Color    Radius    Height   
75              OAK    SPHERE  yellow  0.090000  0.000000  \
48            PAPER    SPHERE   black  0.090000  0.000000   
1          ALUMINUM    SPHERE    blue  0.090000  0.000000   
88  STAINLESS STEEL  CYLINDER  orange  0.100000  0.080000   
8       POLYSTYRENE    SPHERE  orange  0.080000  0.000000   
..              ...       ...     ...       ...       ...   
18  STAINLESS STEEL      CONE   white  0.012192  0.003048   
12            PAPER      CUBE  yellow  0.000000  0.006096   
29  STAINLESS STEEL      CUBE   black  0.000000  0.003048   
17  STAINLESS STEEL      CUBE   black  0.000000  0.003048   
2       POLYSTYRENE      CUBE   black  0.000000  0.003048   

   Units (Imperial/Metric)  Pressure P1  Pressure P2  Piston movement   
75                       M       107451       102581               -1  \
48                       M        96040        99058               -1   
1                        M       100849       105818               -1   
88                       M       101585       103329               -1   
8                        M       104983        96234               -1   
..                     ...          ...          ...              ...   
18                       I        94308        98428                1   
12                       I       103594       103441               -1   
29                       I       101492        93867               -1   
17                       I        93161        99214                1   
2                        I        93328       106262                1   

    delta_p        Volume  
75    -4870  3.053628e-03  
48     3018  3.053628e-03  
1      4969  3.053628e-03  
88     1744  2.513274e-03  
8     -8749  2.144661e-03  
..      ...           ...  
18     4120  4.744533e-07  
12     -153  2.265348e-07  
29    -7625  2.831685e-08  
17     6053  2.831685e-08  
2     12934  2.831685e-08  

[100 rows x 11 columns]

#### A5.
Assuming the definiton of largest object to be the one that has the most volume.
from the above df, we can see that the object with the largest volume is the Oak Sphere. 
The row is 77 in the excel sheet and the index is 75 